runtime/dart_vm_lifecycle.cc

```diff
@@ -0,0 +1,106 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/runtime/dart_vm_lifecycle.h"
+
+#include <mutex>
+
+namespace blink {
+
+// We need to explicitly put the constructor and destructor of the DartVM in the
+// critical section. All accesses (not just const members) to the global VM
+// object weak pointer are behind this mutex.
+static std::mutex gVMMutex;
+static std::weak_ptr<DartVM> gVM FML_GUARDED_BY(gVMMutex);
+
+// We are going to be modifying more than just the control blocks of the
+// following weak pointers (in the |Create| case where an old VM could not be
+// reused). Ideally, we would use |std::atomic<std::weak_ptr<T>>| specialization
+// but that is only available since C++20. We don't expect contention on these
+// locks so we just use one mutex for all.
+static std::mutex gVMDependentsMutex;
+static std::weak_ptr<const DartVMData> gVMData
+    FML_GUARDED_BY(gVMDependentsMutex);
+static std::weak_ptr<ServiceProtocol> gVMServiceProtocol
+    FML_GUARDED_BY(gVMDependentsMutex);
+static std::weak_ptr<IsolateNameServer> gVMIsolateNameServer
+    FML_GUARDED_BY(gVMDependentsMutex);
+
+DartVMRef::DartVMRef(std::shared_ptr<DartVM> vm) : vm_(vm) {}
+
+DartVMRef::DartVMRef(DartVMRef&& other) = default;
+
+DartVMRef::~DartVMRef() {
+  if (!vm_) {
+    // If there is no valid VM (possible via a move), there is no way that the
+    // decrement on the shared pointer can cause a collection. Avoid acquiring
+    // the lifecycle lock in this case. This is just working around a
+    // pessimization and is optinal.
+    return;
+  }
+  std::lock_guard<std::mutex> lifecycle_lock(gVMMutex);
+  vm_.reset();
+}
+
+DartVMRef DartVMRef::Create(Settings settings,
+                            fml::RefPtr<DartSnapshot> vm_snapshot,
+                            fml::RefPtr<DartSnapshot> isolate_snapshot,
+                            fml::RefPtr<DartSnapshot> shared_snapshot) {
+  std::lock_guard<std::mutex> lifecycle_lock(gVMMutex);
+
+  // If there is already a running VM in the process. Grab a strong reference to
+  // it.
+  if (auto vm = gVM.lock()) {
+    FML_LOG(INFO) << "Attempted to create a VM in a process where one was "
+                     "already running. Ignoring arguments for current VM "
+                     "create call and reusing the old VM.";
```

Comment:
nit: Should this be `WARNING`? I'd imagine there shouldn't be many cases where anyone expects they might do an extra create, and calling create with one entrypoint, but getting the previous one feels warn-able.